pilot/pkg/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leaderelection
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/atomic"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	electionID = "istio-leader"
+)
+
+type LeaderElection struct {
```

Comment:
nit: the name creates a stutter with the package name `leaderelection.LeaderElection`. Also, I think `Elector` is a little better than `Election`. Perhaps the package could be `leader` and the type could be `Elector`, so that callsites would see `leader.Elector`? WDYT?